ffmpeg_tools/commands.py

```diff
@@ -449,6 +552,20 @@ def replace_streams_command(input_file,
         for index in subtitle_streams_to_strip
     ]
 
+    subtitle_codec_map = shift_stream_indexes(
+        adjust_stream_indexes_for_removals(
+            select_subtitle_conversions(input_metadata, container),
+            data_streams_to_strip +
+            subtitle_streams_to_strip +
+            meta.find_stream_indexes(input_metadata, VALID_STREAM_TYPES[stream_type]),
+        ),
+        meta.count_streams(replacement_metadata, VALID_STREAM_TYPES[stream_type]),
+    )
+    subtitle_codec_options = [
+        [f"-codec:{index}", codec]
+        for index, codec in subtitle_codec_map.items()
+    ]
```

Comment:
It would indeed make the implementation a bit simpler. But it's not really about the API. I could make it that simple even by just converting the current code into a loop over all the input streams from the two input files (though wrapping it in a class of course has additional benefits).

The reason why it's more complex now is that I tried to keep the ffmpeg command simple. Note for example that before adding subtitles the mappings boiled down to `-map 1:v -map 0 -map -0:v` (take all streams from input `#0` except for video streams and add all video streams from input `#1`). Before we had support for subtitles and data streams it allowed me to perform this operation without having to get metadata for the streams and have ffmpeg do all the dirty work for me.

Obviously, now the command does have to get metadata so what you suggest makes sense. If we already have to get it, we can just as well use it and just loop over it. By the time I had this metadata though I already had this implemented like this so I didn't want to just immediately rewrite it again :)

Having said that, I think such a redesign would be a good thing but I'd rather limit myself to fixing bugs and defects here. Let's finish getting these pull requests through the review and doing any fixes critical to the functionality first. I'm going to fix any stuff that does not work as specified in the tasks we agreed on (and even do some minor tweaks you request as long as that's not too labor-intensive) but if there's going to be more work (like further tasks or refactoring), we should discuss that in the main e-mail thread. And unfortunately my availability in the coming months will be somewhat limited as I have work on other projects queued up (though it really depends on the terms).

As for estimation:
- Doing just a loop sounds like 1 day tops. It would be all contained in the replace function and covered by already existing tests.
- `ReplaceCommandBuilder` class - maybe two or three days, depending on how generic and well tested you'd want it to be. If not very generic then maybe it would be even just 1 day but there's some uncertainty so I don't want to promise that.
- As far as refactoring goes, having something even more general, like a `FfmpegCommand` class that lets you build any of the commands and hides the ugly details inside (like matching inputs to outputs, inserting parameters in the right places in the command (ffmpeg is pretty sensitive to that) detecting contradictions, preventing duplicates, etc.) would be a great idea in my opinion. But that's basically a full rewrite of the `commands` module so significantly more work, especially when you take testing all the corner cases into account.

Reply:
Let's take minimalistic approach than. We will leave it as is.
But QA will test it precisely, because I don't know if we can trust this code.